internal/git/gitsync_processor.go

```diff
@@ -0,0 +1,72 @@
+package git
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1 "github.com/numaproj-labs/numaplane/api/v1"
+)
+
+type Message struct {
+	Updates bool
+	Err     error
+}
+
+type GitSyncProcessor struct {
+	gitSync     v1.GitSync
+	channels    map[string]chan Message
+	k8Client    client.Client
+	clusterName string
+}
+
+func isCommitHash(hash string) bool {
+	match, _ := regexp.MatchString("^[0-9a-fA-F]{40}$", hash)
+	return match
+}
+
+func isValidTag(tag string) bool {
+	match, _ := regexp.MatchString(`^[vV]?\d+\.\d+\.\d+$`, tag)
+	return match
+}
+
+// reference can  can be a branch, a tag, or a commit hash
+func cloneRepository(repoUrl string, reference string) (*git.Repository, error) {
+	var referenceName plumbing.ReferenceName
+	if isValidTag(reference) {
+		referenceName = plumbing.ReferenceName(fmt.Sprintf("refs/tags/%s", reference))
+	} else if isCommitHash(reference) {
+		referenceName = plumbing.NewHashReference("", plumbing.NewHash(reference)).Name()
+	}
```

Comment:
we don't need to set "referenceName" in the case that we don't match either of those first two cases?

Reply:
also, should we allow the case of a commit hash? That should never change, right? And I'm not sure about the case of a tag? Should tags ever change? 

In general, it seems like we want to be monitoring a branch.

Reply:
I guess it seems like ArgoCD does do {commit, tag, branch} based on [this](https://argo-cd.readthedocs.io/en/stable/operator-manual/architecture/#repository-server). But I'm trying to wrap my head around the motivation for monitoring a commit hash or a tag. I'm wondering if we need to continue monitoring in that case or just clone and be done.

Reply:
Per [slack](https://intuit-teams.slack.com/archives/C05K82P3N9J/p1704998292080539) discussion, let's allow branch and tag, not commit.

Reply:
I have removed cloning as per commit hash ,now cloning on the basis of tag and branch @xdevxy @juliev0

Reply:
oh, I think @xdevxy (Hao) is saying above that she does want to allow for cloning a commit hash as well - it's just that in that case the goroutine wouldn't need to continue monitoring anything.

Reply:
So if the user is cloning on commit hash we dnt want to monitor it ? or how it would be

Reply:
right, if it's a branch or a tag, we will continually check if it changed. But if it's a commit, it should never change, so there's no point to continually check. (At the moment, you haven't added any code here to continually check anyway of course.)